### PR TITLE
Feature - Search - Implement API

### PR DESCRIPTION
**Pull request by at-thinhnguyen6**

### CODE REVIEW

File: TheRestaurants/Model/Scheme/SearchHistory.swift [hunk 1]
```diff
@@ -8,7 +8,6 @@
 
 import Foundation
 import RealmSwift
-
```
Comment (at-vanle) [inline TheRestaurants/Model/Scheme/SearchHistory.swift:11]: WARNING - Please insert empty after type declaration.

File: TheRestaurants/View/Search/HistorySearch/HistorySearchCellModel.swift [hunk 2]
```diff
@@ -7,7 +7,6 @@
 //
 
 import Foundation
-
```
Comment (at-vanle) [inline TheRestaurants/View/Search/HistorySearch/HistorySearchCellModel.swift:10]: WARNING - Please insert empty after type declaration.

File: TheRestaurants/View/Search/SearchViewModel.swift [hunk 3]
```diff
@@ -14,22 +14,41 @@ enum Search {
 }
 
 class SearchViewModel {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewModel.swift:16]: WARNING - Please insert empty after type declaration.

File: TheRestaurants/View/Search/SearchViewModel.swift [hunk 4]
```diff
@@ -14,22 +14,41 @@ enum Search {
...
-    var result: [City] = []
+    var results: [Restaurant] = []
     var search: [Search] = [.resultSearch, .historySearch]
     var history: [SearchHistory] = []
 
-    func numberOfRowsInSection(section: Int) -> Int {
-        switch search[section] {
-        case .historySearch:
-            return history.count
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewModel.swift:24]: WARNING - Opening braces should be preceded by a single space and on the same line as the declaration.

File: TheRestaurants/View/Search/ResultSearch/SearchResultCell.swift [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    @IBOutlet weak var ratingLabel: UILabel!
+    @IBOutlet weak var locationLabel: UILabel!
+    @IBOutlet weak var delivery: UILabel!
+
+    var viewModel: SearchResultCellModel? {
+        didSet {
+            setupView()
+        }
+    }
+
```
Comment (at-vanle) [inline TheRestaurants/View/Search/ResultSearch/SearchResultCell.swift:24]: WARNING - Lines should not have trailing whitespace.

File: TheRestaurants/View/Search/ResultSearch/SearchResultCellModel.swift [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+//  TheRestaurants
+//
+//  Created by ADMIN on 9/3/2563 BE.
+//  Copyright © 2563 BE Thinh Nguyen P[6]. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+class SearchResultCellModel {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/ResultSearch/SearchResultCellModel.swift:12]: WARNING - Please insert empty after type declaration.

File: TheRestaurants/View/Search/ResultSearch/SearchResultCellModel.swift [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    var restaurant: Restaurant
+    init(restaurantData: Restaurant) {
+        self.restaurant = restaurantData
+    }
+
```
Comment (at-vanle) [inline TheRestaurants/View/Search/ResultSearch/SearchResultCellModel.swift:17]: WARNING - Lines should not have trailing whitespace.

**Comment by at-vanle** [comment]
Total linter violations in pull request: 7
Serious: 0
Warnings: 7

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 8]
```diff
@@ -50,14 +65,22 @@ extension SearchViewController: UITableViewDataSource {
     }
 
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        switch viewModel.search[indexPath.row] {
+        switch viewModel.search[indexPath.section] {
         case .historySearch:
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:69]: WARNING - Closure expressions should have a single space inside each brace.

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 9]
```diff
@@ -21,27 +21,42 @@ class SearchViewController: BaseViewController {
...
+                this.alert(error: error)
+            }
+        }
+    }
+
     func setUpSearchBar() {
         let searchBar = UISearchBar(frame: CGRect(x: 0, y: 0, width: view.bounds.width - 35, height: 0))
         if let textfield = searchBar.value(forKey: "searchField") as? UITextField {
             textfield.backgroundColor = #colorLiteral(red: 1.0, green: 1.0, blue: 1.0, alpha: 1.0)
             textfield.placeholder = "Tìm kiếm địa điểm "
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:50]: WARNING - Line should be 132 characters or less: currently 170 characters

File: TheRestaurants/View/Search/SearchViewController.swift (hunk 8 above)
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:69]: WARNING - Line should be 132 characters or less: currently 160 characters

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 10]
```diff
@@ -50,14 +65,22 @@ extension SearchViewController: UITableViewDataSource {
     }
 
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        switch viewModel.search[indexPath.row] {
+        switch viewModel.search[indexPath.section] {
         case .historySearch:
-            guard let cell = tableView.dequeueReusableCell(withIdentifier: "historycell", for: indexPath) as? HistorySearchCell else { return UITableViewCell ()}
-            cell.viewModel = viewModel.viewModelForCell(at: indexPath)
+            guard let cell = tableView.dequeueReusableCell(withIdentifier: "historycell", for: indexPath) as? HistorySearchCell else { return UITableViewCell()}
+            cell.viewModel = viewModel.viewModelForCellHistoried(at: indexPath)
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:71]: WARNING - Line should be 132 characters or less: currently 151 characters

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 11]
```diff
@@ -16,48 +16,109 @@ class SearchViewController: BaseViewController {
...
     func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return viewModel.numberOfRowsInSection(section: section)
+        if tableView == listHistoriedSerch {
+            return viewModel.numberOfRowInSectionHistoried()
+        }
+        return viewModel.numberOfRowInSectionResult()
     }
 
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        switch viewModel.search[indexPath.row] {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:53]: WARNING - Line should be 132 characters or less: currently 147 characters

**Comment by at-vanle** [comment]
Total linter violations in pull request: 18
Serious: 0
Warnings: 18

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 12]
```diff
@@ -35,29 +63,88 @@ class SearchViewController: BaseViewController {
...
 }
-extension SearchViewController: UITableViewDataSource {
+extension SearchViewController: UITableViewDataSource, UITableViewDelegate {
     func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return viewModel.numberOfRowsInSection(section: section)
+        if tableView == listHistoriedSerch {
+            return viewModel.numberOfRowInSectionHistoried()
+        }
+        return viewModel.numberOfRowInSectionResult()
     }
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:97]: WARNING - Line should be 132 characters or less: currently 134 characters

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 13]
```diff
@@ -16,48 +16,109 @@ class SearchViewController: BaseViewController {
...
 
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        switch viewModel.search[indexPath.row] {
-        case .historySearch:
-            guard let cell = tableView.dequeueReusableCell(withIdentifier: "historycell", for: indexPath) as? HistorySearchCell else { return UITableViewCell ()}
-            cell.viewModel = viewModel.viewModelForCell(at: indexPath)
-            return cell
-        case .resultSearch:
-            guard let cell = tableView.dequeueReusableCell(withIdentifier: "searchcell", for: indexPath) as? HomeCell else { return UITableViewCell() }
+        if tableView == listResultSearch {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift]: WARNING - Line should be 132 characters or less: currently 134 characters

File: TheRestaurants/View/Search/SearchViewModel.swift [hunk 14]
```diff
@@ -7,29 +7,93 @@
...
+            print(error)
+        }
+    }
+
+    func addRealm(searchKey: String, completion: @escaping (Bool) -> Void) {
+        do {
+            let realm = try Realm()
+            let result = SearchHistory()
+            result.searchKey = searchKey
+            try realm.write {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewModel.swift:86]: WARNING - Must have an empty line between functions.

**Comment by at-vanle** [comment]
Total linter violations in pull request: 13
Serious: 0
Warnings: 13

**Comment by at-vanle** [comment]
Total linter violations in pull request: 12
Serious: 0
Warnings: 12

File: TheRestaurants/Model/API/Api.Search.swift [hunk 15]
```diff
@@ -39,4 +39,28 @@ extension Api.Search {
...
+            DispatchQueue.main.async {
+                switch result {
+                case .success(let value):
+                    guard let value = value as? JSObject, let restaurants = value["restaurants"] as? JSArray else {
+                        completion(.failure(Api.Error.json))
+                        return
+                    }
+                    var results: [Restaurant] = []
+                    for item in restaurants {
+                        guard let restaurant = item["restaurant"] as? JSObject,
```
Comment (at-vanle) [inline TheRestaurants/Model/API/Api.Search.swift:55]:
```suggestion
                        guard let json = item["restaurant"] as? JSObject,
```

File: TheRestaurants/Model/API/Api.Search.swift [hunk 16]
```diff
@@ -39,4 +39,28 @@ extension Api.Search {
...
+                        let restaurant2 = Mapper<Restaurant>().map(JSONObject: restaurant) else { return }
```
Comment (at-vanle) [inline TheRestaurants/Model/API/Api.Search.swift:56]:
```suggestion
                        let restaurant = Mapper<Restaurant>().map(JSONObject: restaurant) else { return }
```

File: TheRestaurants/Model/Scheme/Restaurant.swift [hunk 17]
```diff
@@ -8,6 +8,7 @@
 
 import Foundation
 import ObjectMapper
+import RealmSwift
```
Comment (at-vanle) [inline TheRestaurants/Model/Scheme/Restaurant.swift:11]: Import mà không dùng à?

File: TheRestaurants/View/Search/ResultSearch/SearchResultCell.swift [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+//  SearchResultCell.swift
+//  TheRestaurants
+//
+//  Created by ADMIN on 9/3/2563 BE.
+//  Copyright © 2563 BE Thinh Nguyen P[6]. All rights reserved.
+//
+
+import UIKit
+
+class SearchResultCell: UITableViewCell {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/ResultSearch/SearchResultCell.swift:11]: Đã nói là dùng Home Cell rồi mà?

File: TheRestaurants/View/Search/ResultSearch/SearchResultCellModel.swift [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+//  TheRestaurants
+//
+//  Created by ADMIN on 9/3/2563 BE.
+//  Copyright © 2563 BE Thinh Nguyen P[6]. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+class SearchResultCellModel {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/ResultSearch/SearchResultCellModel.swift:12]: Dùng HomeCell

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 20]
```diff
@@ -12,21 +12,51 @@ class SearchViewController: BaseViewController {
...
+
     override func viewDidLoad() {
         super.viewDidLoad()
         setUpSearchBar()
         configTableView()
+        fetchData()
     }
 
-    var viewModel = SearchViewModel()
+    private func addRealm(searchKey: String) {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:25]:
```suggestion
    private func saveKeyToRealm(searchKey: String) {
```

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 21]
```diff
@@ -12,21 +12,51 @@ class SearchViewController: BaseViewController {
...
+        viewModel.addRealm(searchKey: searchKey) {  [weak self] (result) in
+            guard let this = self else { return }
+            switch result {
+            case.success:
+                this.fetchData()
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:30]: Gì đây? Để làm gì?

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 22]
```diff
@@ -12,21 +12,51 @@ class SearchViewController: BaseViewController {
 
     @IBOutlet private weak var listHistoriedSerch: UITableView!
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:13]:
```suggestion
    @IBOutlet private weak var searchHistoryTableView: UITableView!
```

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 23]
```diff
@@ -12,21 +12,51 @@ class SearchViewController: BaseViewController {
...
     @IBOutlet private weak var listResultSearch: UITableView!
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:14]:
```suggestion
    @IBOutlet private weak var resultTableView: UITableView!
```

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 24]
```diff
@@ -35,29 +65,89 @@ class SearchViewController: BaseViewController {
...
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        if tableView == listResultSearch {
+            let viewController = DetailViewController()
+            viewController.viewModel = viewModel.didSelectRowAt(indexPath: indexPath)
+            viewController.hidesBottomBarWhenPushed = true
+            navigationController?.pushViewController(viewController, animated: true)
+        } else {
+            let keywork = viewModel.histories[indexPath.row].searchKey
+            loadCell(keyword: keywork)
+            listResultSearch.isHidden = false
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:125]: Cái này phải xử lý trong completion của hàm **loadCell**

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 25]
```diff
@@ -35,29 +65,89 @@ class SearchViewController: BaseViewController {
...
+            listHistoriedSerch.isHidden = true
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:126]: Như trên

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 26]
```diff
@@ -35,29 +65,89 @@ class SearchViewController: BaseViewController {
...
+        listResultSearch.isHidden = true
+        fetchData()
+    }
+
+    func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
+        listHistoriedSerch.isHidden = true
+        listResultSearch.isHidden = false
+    }
+
+    func searchBarShouldEndEditing(_ searchBar: UISearchBar) -> Bool {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:150]: Làm gì?

File: TheRestaurants/View/Search/SearchViewModel.swift [hunk 27]
```diff
@@ -7,29 +7,81 @@
...
-    var history: [SearchHistory] = []
-
-    func numberOfRowsInSection(section: Int) -> Int {
-        switch search[section] {
-        case .historySearch:
-            return history.count
-        case .resultSearch:
-            return result.count
+    var histories: [SearchHistory] = []
+    private var notificationToken: NotificationToken?
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewModel.swift:21]: Làm gì đây?

File: TheRestaurants/View/Search/SearchViewModel.swift [hunk 28]
```diff
@@ -7,29 +7,81 @@
...
 
 enum Search {
     case resultSearch
     case historySearch
 }
 
 class SearchViewModel {
-    var result: [City] = []
+    var results: [Restaurant] = []
     var search: [Search] = [.resultSearch, .historySearch]
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewModel.swift:19]: Không thấy dùng

File: TheRestaurants/View/Search/SearchViewModel.swift [hunk 29]
```diff
@@ -7,29 +7,81 @@
...
+                completion( .success)
+            case .failure(let error):
+                completion(.failure(error))
+            }
         }
     }
 
-    func viewModelForCell(at indexPath: IndexPath) -> HistorySearchCellModel {
-        let item = history[indexPath.row]
+    func viewModelForCellHistoried(at indexPath: IndexPath) -> HistorySearchCellModel {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewModel.swift:45]:
```suggestion
    func viewModelForHistoryCell(at indexPath: IndexPath) -> HistorySearchCellModel {
```

File: TheRestaurants/View/Search/SearchViewModel.swift [hunk 30]
```diff
@@ -7,29 +7,81 @@
...
+        let item = histories[indexPath.row]
         let viewModel = HistorySearchCellModel(searchHistory: item)
         return viewModel
     }
+
+    func viewModelForCellResult(indexPath: IndexPath) -> HomeCellModel {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewModel.swift:51]:
```suggestion
    func viewModelForResultCell(indexPath: IndexPath) -> HomeCellModel {
```

File: TheRestaurants/View/Search/SearchViewModel.swift [hunk 31]
```diff
@@ -7,29 +7,81 @@
...
+        let item = results[indexPath.row]
+        let viewModel = HomeCellModel(cellsRestaurant: item)
+        return viewModel
+    }
+
+    func fetchData(completion: @escaping APICompletion) {
+        do {
+            let realm = try Realm()
+            let results = realm.objects(SearchHistory.self)
+            histories = Array((results).reversed().prefix(7))
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewModel.swift:61]: Chỗ này xử lý sao đây? Vì sao phải **prefix(7)**?

File: TheRestaurants/View/Search/SearchViewModel.swift [hunk 32]
```diff
@@ -7,29 +7,81 @@
...
+            completion(.success)
+        } catch {
+            print(error)
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewModel.swift:64]: **completion(.failure(error))**

File: TheRestaurants/View/Search/SearchViewModel.swift [hunk 33]
```diff
@@ -7,29 +7,81 @@
...
         return viewModel
     }
+
+    func viewModelForCellResult(indexPath: IndexPath) -> HomeCellModel {
+        let item = results[indexPath.row]
+        let viewModel = HomeCellModel(cellsRestaurant: item)
+        return viewModel
+    }
+
+    func fetchData(completion: @escaping APICompletion) {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewModel.swift:57]:
```suggestion
    func fetchSearchHistoryData(completion: @escaping APICompletion) {
```

File: TheRestaurants/View/Search/SearchViewModel.swift [hunk 34]
```diff
@@ -7,29 +7,81 @@
...
+            let realm = try Realm()
+            let results = realm.objects(SearchHistory.self)
+            histories = Array((results).reversed().prefix(7))
+            completion(.success)
+        } catch {
+            print(error)
+        }
+    }
+
+    func addRealm(searchKey: String, completion: @escaping APICompletion) {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewModel.swift:68]:
```suggestion
    func saveKeyToRealm(searchKey: String, completion: @escaping APICompletion) {
```

**Review by at-vanle — CHANGES_REQUESTED** [review]
@at-thinhnguyen6 Please check my comments. Thanks!

File: TheRestaurants/View/Search/SearchViewModel.swift (hunk 31 above)
Comment (at-thinhnguyen6) [reply TheRestaurants/View/Search/SearchViewModel.swift:61]: dạ e muốn lấy 7 cái đầu tiên cho gọn ạ

**Comment by at-vanle** [comment]
Total linter violations in pull request: 13
Serious: 0
Warnings: 13

File: TheRestaurants/View/Search/SearchViewController.swift [hunk 35]
```diff
@@ -35,29 +67,83 @@ class SearchViewController: BaseViewController {
...
                 leftView.image = leftView.image?.withRenderingMode(.alwaysTemplate)
                 leftView.tintColor = #colorLiteral(red: 0, green: 0, blue: 0, alpha: 0.5125578704)
             }
         }
-         navigationItem.rightBarButtonItem = UIBarButtonItem(customView: searchBar)
+        searchBar.delegate = self 
+        navigationItem.rightBarButtonItem = UIBarButtonItem(customView: searchBar)
+    }
+
+    func fetchData() {
```
Comment (at-vanle) [inline TheRestaurants/View/Search/SearchViewController.swift:79]:
```suggestion
    func fetchSearchHistoryData() {
```

**Review by at-vanle — CHANGES_REQUESTED** [review]
@at-thinhnguyen6 Please check my comments. Thank you!

**Review by at-vanle — APPROVED** [review]
LGTM!

**Pull request merged**